contracts/src/v0.8/functions/dev/v1_X/libraries/ChainSpecificUtil.sol

```diff
@@ -40,18 +40,24 @@ library ChainSpecificUtil {
 
   // ------------ End Optimism Constants ------------
 
-  /// @notice Returns the L1 fees in wei that will be paid for the current transaction, given any calldata
-  /// @notice for the current transaction.
-  /// @notice When on a known Arbitrum chain, it uses ArbGas.getCurrentTxL1GasFees to get the fees.
-  /// @notice On Arbitrum, the provided calldata is not used to calculate the fees.
-  /// @notice On Optimism, the provided calldata is passed to the GasPriceOracle predeploy
-  /// @notice and getL1Fee is called to get the fees.
-  function _getCurrentTxL1GasFees(bytes memory txCallData) internal view returns (uint256 l1FeeWei) {
+  /// @notice Returns the upper limit estimate of the L1 fees in wei that will be paid for L2 chains
+  /// @notice based on the size of the transaction data and the current gas conditions.
+  /// @notice This is an "upper limit" as it assumes the transaction data is uncompressed when posted on L1.
+  function _getL1FeeUpperLimit(uint256 dataSizeBytes) internal view returns (uint256 l1FeeWei) {
     uint256 chainid = block.chainid;
     if (_isArbitrumChainId(chainid)) {
-      return ARBGAS.getCurrentTxL1GasFees();
+      // https://docs.arbitrum.io/build-decentralized-apps/how-to-estimate-gas#where-do-we-get-all-this-information-from
+      (, uint256 l1PricePerByte, , , , ) = ARBGAS.getPricesInWei();
+      return l1PricePerByte * (dataSizeBytes + ARB_L1_FEE_DATA_PADDING_SIZE);
     } else if (_isOptimismChainId(chainid)) {
-      return OVM_GASPRICEORACLE.getL1Fee(bytes.concat(txCallData, L1_FEE_DATA_PADDING));
+      // https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/exec-engine.md#ecotone-l1-cost-fee-changes-eip-4844-da
+      // note we conservatively assume all non-zero bytes: tx_compressed_size = tx_data_size_bytes
+      uint256 l1BaseFeeWei = L1BLOCK.basefee();
+      uint256 l1BaseFeeScalar = L1BLOCK.baseFeeScalar();
+      uint256 l1BlobBaseFeeWei = L1BLOCK.blobBaseFee();
+      uint256 l1BlobBaseFeeScalar = L1BLOCK.blobBaseFeeScalar();
+      uint256 weightedGasPrice = (l1BaseFeeScalar * l1BaseFeeWei + l1BlobBaseFeeScalar * l1BlobBaseFeeWei) / 10 ** 6;
```

Comment:
the formula is `16*base_fee_scalar*base_fee + blob_base_fee_scalar*blob_base_fee`, might be missing the `16*` here

Reply:
There is some nuance here that has to do with compression.
Technically they multiply (4 * zeroBytesCount + 16 * nonzeroBytesCount) / 16
However here we assume all data is uncompressed to get the "upper limit", therefore
(16 * nonzeroBytesCount) / 16 == nonzeroBytesCount
so we shouldn't need the 16*

Can you confirm that this makes sense @matYang ?

Reply:
I mean the 16 is not on the entire gas price, it is on the base fee (calldata) portion and not on the blob portion right?

I I get it that you want to assume every byte is non-zero and have `tx_compressed_size  = dataSizeBytes` to cap the worst case, then the formula becomes

```
l1_data_fee = dataSizeBytes * weighted_gas_price
= dataSizeBytes * ( (16*base_fee_scalar*base_fee) + blob_base_fee_scalar*blob_base_fee)
```